app/models/language.rb

```diff
@@ -58,7 +58,7 @@ def self.submission_options
 
   # for the selection dropdown
   def self.grouped_submission_options
-    current = ["c++", "c", "python", "java", "haskell", "ruby", "j"]  # language group identifiers, order is preserved
+    current = ["c++", "csharp", "c", "python", "java", "haskell", "ruby", "j"]  # language group identifiers, order is preserved
```

Comment:
```suggestion
    current = ["c++", "c", "python", "csharp", "java", "haskell", "ruby", "j"]  # language group identifiers, order is preserved
```
The "Current" languages are ordered by preference; I think C# should be further down (after Python).

Reply:
In that case, should we also move Python above C?

Reply:
I feel like C and C++ should be together...

Reply:
We should probably decide what to order on. If it is by "preference" then Python is more popular for NZIC followed by C++. Hardly anyone uses C so, although I agree that they usually belong together, it would make more sense under this scheme to have C further down.

Reply:
But it seems, at least, we all agree that Python should move ahead of C#. So that's something :)

Reply:
Yes, I think everyone loves Python :)

I'm still not sure about moving C down. I was going to post the following as a separate comment, but it seems like it might be somewhat relevant to the current discussion:

> The language list is getting pretty long (bearing in mind we'll also add JavaScript), so searching will be difficult again (previous discussion: https://github.com/NZOI/nztrain/pull/52#discussion_r368271778). I'm starting to think that @Holmes98's original idea to sort the entire list might have been the right thing to do all along. There is no need to hold up this PR, but we should think about it. Some ideas:
>  - Move J to "Other" (or remove it completely; only Ronald has every used it)
>  - Move Ruby to "Other"
>  - Remove old versions: C99 (as @Holmes98 originally wanted to do), C++11 (we should keep C++14 for the IOI, but I can't think of a reason to keep C++11). This will help if we sort the entire list.

Reply:
I'll later submit a PR that will set each user's default language to the language of their latest submission. For almost everyone this should remove the need to re-select their language on each submission, so I honestly don't really mind what order we have the languages in. The only issue is that a lot of people will then have their default as C++11 instead of C++17 (and might not notice the new version) so I was thinking of also removing C++11 at the same time.

Reply:
I second removing J.

Perhaps a better name for the "Other" group is "Older" or "Outdated"?

Since Python and C++ are the most popular by a mile, maybe we put Python and C++ first then sort the remaining languages in "Current" alphabetically. If we put Python first then C++ and C would conveniently appear next to each other.

Reply:
Otherwise, sorting each group would mean we don't have to keep having this discussion every time a new language is added. But I think there is value in separating out the outdated Ruby and C++14/11 versions while we have them

Reply:
> I'll later submit a PR that will set each user's default language to the language of their latest submission.

Sounds great. The issue of not noticing when a newer version becomes available is annoying though, and will come up again every time we update the language.

Would it be too complicated/slow to add some logic along the lines of

"If the last submission's language is no longer the current version within its group, and the user has never submitted in the current version of the language, then use the current version of that language as the default."

This is meant to catch cases where a user tries C++17 but then intentionally wants to go back to C++14. But if a user is careful to never select C++17, then it will keep trying to set it as their default. The only solution I can think of is to also record the date when the language version was added, which seems too much hassle.

> Perhaps a better name for the "Other" group is "Older" or "Outdated"? In which case, Ruby fits (for now).

I thought that at first, but I started liking "Other" because it gave us the freedom to put less-preferred (but still up-to-date) languages there. So we could make C++ and Python into the "Current" languages, and everything else into "Other". (But then maybe rename "Current" to "Favourites" or something.)

Reply:
> I thought that at first, but I started liking "Other" because it gave us the freedom to put less-preferred (but still up-to-date) languages there.

How would we define "preferred"? In terms of the IOI and camp the only language we prefer is C++. The fact that we all like Python is favouritism on our part

Reply:
To throw another idea into the pop, implement a time relevant frequency counter that counts the number of submissions in each language upon submission. We would then lookup each language by frequency counter value and display then in that order. One issue here, aside from implementation, could be that the most popular language will oscillate on a yearly basis (Python by the end of NZIC and C++ by the end of camp).

Reply:
> How would we define "preferred"?

I was thinking it would be to do with what we teach in camp (mainly C++, but the NZIC editorials use Python) but also what students commonly use.

> lookup each language by frequency counter

A hard-coded list of favourite languages is like a cheap version of this; we can manually run some queries on the database if we want to see which languages are popular, but then we can also factor in the camp's bias ("C++ first").

Reply:
Would it make sense then to have C++ and Python as the NZIC official languages and use three groups in the menu...

Official
    C++
    Python

Current
    [list alphabetically common and up-to-date languages]

Other
    [Old versions + random languages]

Reply:
How about we merge the last two?
```
Favourite/Pinned/Official
    C++
    Python
Other
    [list everything alphabetically]
```

Reply:
(if we can't settle on good names, we can use a separator, "----")

Reply:
I think we couldn't work out how to add a non-selectable separator (without optgroups) but otherwise I like the last option the most.

Reply:
Agreed. The non-selectable separator was before I started using groups, so hopefully it shouldn't be difficult.

Reply:
This makes a non-selectable separator: `["────────", "", {disabled: true}]`
(credit for the unicode magic: https://stackoverflow.com/questions/899148/html-select-option-separator)